core/chains/evm/headtracker/head_saver_in_mem.go

```diff
@@ -0,0 +1,163 @@
+		Heads:       make(map[BLOCK_HASH]H),
+		HeadsNumber: make(map[int64][]H),
+		getNilHead:  getNilHead,
+		setParent:   setParent,
+	}
+}
+
+func NewEvmInMemoryHeadSaver(config Config, lggr logger.Logger) *EvmInMemoryHeadSaver {
+	evmConfig := NewWrappedConfig(config)
+	return NewInMemoryHeadSaver[*evmtypes.Head, common.Hash, *big.Int](
+		evmConfig,
+		lggr,
+		func() *evmtypes.Head { return nil },
+		func(head, parent *evmtypes.Head) { head.Parent = parent },
+	)
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) Save(ctx context.Context, head H) error {
+	if !head.IsValid() {
+		return nil
+	}
+
+	historyDepth := int64(hs.config.HeadTrackerHistoryDepth())
+	hs.AddHeads(historyDepth, head)
+
+	return nil
+}
+
+// No OP function for EVM
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) Load(ctx context.Context) (H, error) {
+
+	return hs.LatestChain(), nil
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) LatestChain() H {
+	head := hs.getLatestHead()
+
+	if head.ChainLength() < hs.config.FinalityDepth() {
+		hs.logger.Debugw("chain shorter than EvmFinalityDepth", "chainLen", head.ChainLength(), "evmFinalityDepth", hs.config.FinalityDepth())
+	}
+	return head
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) Chain(blockHash BLOCK_HASH) H {
+	hs.mu.RLock()
+	defer hs.mu.RUnlock()
+
+	if head, exists := hs.Heads[blockHash]; exists {
+		return head
+	}
+
+	return hs.getNilHead()
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) HeadByNumber(blockNumber int64) []H {
+	hs.mu.RLock()
+	defer hs.mu.RUnlock()
+
+	return hs.HeadsNumber[blockNumber]
+}
+
+// Assembles the heads together and populates the Heads Map
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) AddHeads(historyDepth int64, newHeads ...H) {
+	hs.mu.Lock()
+	defer hs.mu.Unlock()
+
+	hs.trimHeads(historyDepth)
+
+	for _, head := range newHeads {
+		blockHash := head.BlockHash()
+		blockNumber := head.BlockNumber()
+
+		if _, exists := hs.Heads[blockHash]; exists {
+			continue
+		}
+
+		if parent, exists := hs.Heads[blockHash]; exists {
+			hs.setParent(head, parent)
+		} else {
+			// If parent's head is too old, we should set it to nil
+			hs.setParent(head, hs.getNilHead())
+		}
+
+		hs.Heads[blockHash] = head
+		hs.HeadsNumber[blockNumber] = append(hs.HeadsNumber[blockNumber], head)
+
+		if !hs.latestHead.IsValid() {
+			hs.latestHead = head
+		} else if head.BlockNumber() > hs.latestHead.BlockNumber() {
+			hs.latestHead = head
+		}
+	}
+}
+
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) TrimOldHeads(historyDepth int64) {
+	hs.mu.Lock()
+	defer hs.mu.Unlock()
+
+	hs.trimHeads(historyDepth)
+}
+
+// trimHeads() is should only be called by functions with mutex locking.
+// trimHeads() is an internal function without locking to prevent deadlocks
+func (hs *inMemoryHeadSaver[H, BLOCK_HASH, CHAIN_ID]) trimHeads(historyDepth int64) {
+	for headNumber, headNumberList := range hs.HeadsNumber {
+		if headNumber < historyDepth {
```

Comment:
If historyDepth is just a number indicating the depth to keep, then this needs to be:
if hs.latestHead.Number - headNumber > historyDepth

Reply:
The code here assumes historyDepth = the oldest head number that we need to keep